model/Worker/WorkerProcessManager.php

```diff
@@ -0,0 +1,170 @@
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2017 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ *
+ */
+
+namespace oat\taoTaskQueue\model\Worker;
+
+use Exception;
+use oat\oatbox\log\LoggerAwareTrait;
+use oat\oatbox\service\ConfigurableService;
+use React\ChildProcess\Process;
+
+class WorkerProcessManager extends ConfigurableService
+{
+    use LoggerAwareTrait;
+
+    const SERVICE_ID = 'taoTaskQueue/WorkerProcessManager';
+
+    const OPTION_TASK_COMMAND = 'task_command';
+
+    /** @var Process[] */
+    private $processes;
+
+    /** @var integer */
+    private $limitOfCpu;
+
+    /** @var integer */
+    private $limitOfMemory;
+
+    /**
+     * @param Process $process
+     */
+    public function addProcess(Process $process)
+    {
+        $pid = $process->getPid();
+
+        $process->stdout->on('data', function ($status) use ($pid)  {
+            $this->logInfo('Process: '. $pid .' status:'. $status);
+        });
+
+        $process->stdout->on('end', function () use ($pid)  {
+            $this->logInfo('Process: '. $pid .' ended');
+        });
+
+        $process->stdout->on('error', function (Exception $e) use ($pid)  {
+            $this->logError('Process: '. $pid .' error. ' . $e->getMessage());
+        });
+
+        $process->stdout->on('close', function () use ($pid)  {
+            $this->logInfo('Process: '. $pid .' closed.');
+
+            unset($this->processes[$pid]);
+        });
+
+        $process->stdin->end($data = null);
+
+        $this->processes[$pid] = $process;
+    }
+
+    /**
+     * @return bool
+     */
+    public function canRun()
+    {
+        $this->logInfo('No of process workers running: '. count($this->processes));
+
+        $memoryUsage = $this->getMemoryUsage();
+        $cpuUsage    = $this->getCpuUsage();
+
+        if ($memoryUsage < $this->limitOfMemory
+            && $cpuUsage < $this->limitOfCpu
+        ) {
+            return true;
+        }
+
+        $this->logInfo('Limit Of memory and Cpu exceeded waiting for task to finish.
+        Current memory usage:'.$memoryUsage.' Cpu usage:'.$cpuUsage);
+
+        return false;
+    }
+
+    /**
+     * @return string
+     */
+    public function getCommand()
+    {
+        return $this->getOption(static::OPTION_TASK_COMMAND);
+    }
+
+    /**
+     * @return float|int
+     */
+    public function getMemoryUsage()
+    {
+        $free = shell_exec('free');
```

Comment:
Would be nice to check before if the `free` command is available (see on Windows)

Reply:
Windows it's not a priority, we don't have servers on windows, and for developers on windows the RunTask script can be used.

Reply:
In that case put it in the doc because I'm sure we will have complaints from developers using Windows. ;)